File: src/main/java/de/espend/idea/php/annotation/completion/PhpAnnotationTypeCompletionProvider.java
package de.espend.idea.php.annotation.completion;

import com.intellij.codeInsight.lookup.LookupElementBuilder;
import com.jetbrains.php.lang.documentation.phpdoc.psi.PhpDocComment;
import com.jetbrains.php.lang.documentation.phpdoc.psi.tags.PhpDocTag;
import com.jetbrains.php.lang.psi.elements.Field;
import com.jetbrains.php.lang.psi.elements.PhpPsiElement;
import de.espend.idea.php.annotation.dict.AnnotationPropertyEnum;
import de.espend.idea.php.annotation.extension.PhpAnnotationCompletionProvider;
import de.espend.idea.php.annotation.extension.parameter.AnnotationCompletionProviderParameter;
import de.espend.idea.php.annotation.extension.parameter.AnnotationPropertyParameter;
import de.espend.idea.php.annotation.util.AnnotationUtil;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class PhpAnnotationTypeCompletionProvider implements PhpAnnotationCompletionProvider {

    @Override
    public void getPropertyValueCompletions(AnnotationPropertyParameter annotationPropertyParameter, AnnotationCompletionProviderParameter completionParameter) {
        String propertyName = annotationPropertyParameter.getPropertyName();
        if(annotationPropertyParameter.getType() != AnnotationPropertyParameter.Type.PROPERTY_VALUE || propertyName == null) {
            return;
        }

        Set<String> values = new HashSet<>();
        AnnotationUtil.visitAttributes(annotationPropertyParameter.getPhpClass(), (attributeName, type, target) -> {
            if(!attributeName.equals(propertyName)) {
                return null;
            }

            if (AnnotationPropertyEnum.fromString(type) == AnnotationPropertyEnum.BOOLEAN) {
                values.addAll(Arrays.asList("false", "true"));
            }

            // @Enum({"AUTO", "SEQUENCE"})
            if (target instanceof Field) {
                PhpDocComment docComment = ((Field) target).getDocComment();
                if(docComment != null) {
                    PhpDocTag[] phpDocTags = docComment.getTagElementsByName("@Enum");
                    for(PhpDocTag phpDocTag: phpDocTags) {
                        PhpPsiElement phpDocAttrList = phpDocTag.getFirstPsiChild();
                        if(phpDocAttrList != null) {
                            String enumArrayString = phpDocAttrList.getText();
                            Pattern targetPattern = Pattern.compile("\"(\\w+)\"");

                            Matcher matcher = targetPattern.matcher(enumArrayString);
                            while (matcher.find()) {
                                values.add(matcher.group(1));
                            }
                        }
                    }
                }
            }

            return null;
        });

        for(String s: values) {
            completionParameter.getResult().addElement(LookupElementBuilder.create(s));
        }
    }
}

File: src/main/kotlin/de/espend/idea/php/annotation/completion/PhpAnnotationTypeCompletionProvider.kt
package de.espend.idea.php.annotation.completion

import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.jetbrains.php.lang.psi.elements.Field
import de.espend.idea.php.annotation.dict.AnnotationPropertyEnum
import de.espend.idea.php.annotation.extension.PhpAnnotationCompletionProvider
import de.espend.idea.php.annotation.extension.parameter.AnnotationCompletionProviderParameter
import de.espend.idea.php.annotation.extension.parameter.AnnotationPropertyParameter
import de.espend.idea.php.annotation.util.AnnotationUtil
import java.util.regex.Pattern

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
class PhpAnnotationTypeCompletionProvider : PhpAnnotationCompletionProvider {
    override fun getPropertyValueCompletions(
        annotationPropertyParameter: AnnotationPropertyParameter,
        completionParameter: AnnotationCompletionProviderParameter
    ) {
        val propertyName = annotationPropertyParameter.propertyName
        if (annotationPropertyParameter.type != AnnotationPropertyParameter.Type.PROPERTY_VALUE || propertyName == null) {
            return
        }

        val values = HashSet<String>()
        AnnotationUtil.visitAttributes(
            annotationPropertyParameter.phpClass,
        ) { attributeName, type, target ->
            if (attributeName == propertyName) {
                if (AnnotationPropertyEnum.fromString(type) == AnnotationPropertyEnum.BOOLEAN) {
                    values.addAll(listOf("false", "true"))
                }

                // @Enum({"AUTO", "SEQUENCE"})
                if (target is Field) {
                    val docComment = target.docComment
                    if (docComment != null) {
                        val phpDocTags = docComment.getTagElementsByName("@Enum")
                        for (phpDocTag in phpDocTags) {
                            val phpDocAttrList = phpDocTag.firstPsiChild
                            if (phpDocAttrList != null) {
                                val enumArrayString = phpDocAttrList.text
                                val targetPattern = Pattern.compile("\"(\\w+)\"")

                                val matcher = targetPattern.matcher(enumArrayString)
                                while (matcher.find()) {
                                    values.add(matcher.group(1))
                                }
                            }
                        }
                    }
                }
            }
            null
        }

        for (s in values) {
            completionParameter.result.addElement(LookupElementBuilder.create(s))
        }
    }
}
